inc/reminder.class.php

```diff
@@ -180,84 +180,164 @@ public function haveVisibilityAccess() {
+            if (count($restrict)) {
+               $or = $or + $restrict;
+            }
+            $where['OR'][] = ['AND' => [
+               'glpi_groups_reminders.groups_id' => count($_SESSION["glpigroups"])
+                                                         ? $_SESSION["glpigroups"]
+                                                         : [-1],
+               'OR' => $or
+            ]];
+         }
       }
 
       // Profiles
-      if (isset($_SESSION["glpiactiveprofile"]) && isset($_SESSION["glpiactiveprofile"]['id'])) {
-         $restrict .= " OR (`glpi_profiles_reminders`.`profiles_id`
-                                 = '".$_SESSION["glpiactiveprofile"]['id']."'
-                            AND (`glpi_profiles_reminders`.`entities_id` < 0
-                                 ".getEntitiesRestrictRequest("OR", "glpi_profiles_reminders", '',
-                                                              '', true).")) ";
+      if ($forceall
+          || (isset($_SESSION["glpiactiveprofile"])
+              && isset($_SESSION["glpiactiveprofile"]['id']))) {
+         $join['glpi_profiles_reminders'] = [
+            'FKEY' => [
+               'glpi_profiles_reminders'  => 'reminders_id',
+               'glpi_reminders'           => 'id'
+            ]
+         ];
+
+         if (Session::getLoginUserID()) {
+            $or = ['glpi_profiles_reminders.entities_id' => ['<', 0]];
+            $restrict = getEntitiesRestrictCriteria('glpi_profiles_reminders', '', '', true);
+            if (count($restrict)) {
+               $or = $or + $restrict;
+            }
+            $where['OR'][] = ['AND' => [
+               'glpi_profiles_reminders.profiles_id' => $_SESSION["glpiactiveprofile"]['id'],
+               'OR' => $or
+            ]];
+         }
       }
 
       // Entities
+      if ($forceall
+          || (isset($_SESSION["glpiactiveentities"]) && count($_SESSION["glpiactiveentities"]))) {
+         $join['glpi_entities_reminders'] = [
+            'FKEY' => [
+               'glpi_entities_reminders'  => 'reminders_id',
+               'glpi_reminders'           => 'id'
+            ]
+         ];
+      }
       if (isset($_SESSION["glpiactiveentities"]) && count($_SESSION["glpiactiveentities"])) {
-         // Force complete SQL not summary when access to all entities
-         $restrict .= getEntitiesRestrictRequest("OR", "glpi_entities_reminders", '', '', true, true);
+         if (Session::getLoginUserID()) {
+            $restrict = getEntitiesRestrictCriteria('glpi_entities_reminders', '', '', true, true);
+            if (count($restrict)) {
+               $where['OR'] = $where['OR'] + $restrict;
+            } else {
+               $where['glpi_entities_reminders.entities_id'] = null;
+            }
+         }
       }
 
-      return '('.$restrict.')';
-   }
+      $criteria = ['LEFT JOIN' => $join];
+      if (count($where)) {
+         $criteria['WHERE'] = $where;
```

Comment:
Maybe you should always return a `WHERE` key, even if it is an empty array. It could prevent problems on result usage.